client/report.go

```diff
@@ -9,16 +9,12 @@ import (
 const reportEndpoint = "/transactions.xml"
 
 //ReportAppID - Report for the Application Id authentication pattern with serviceToken
-func (client *ThreeScaleClient) ReportAppID(appId string, token string, tokenType string, serviceId string, transactions ReportTransactions) (ApiResponse, error) {
+func (client *ThreeScaleClient) ReportAppID(appId string, auth TokenAuth, serviceId string, transactions ReportTransactions) (ApiResponse, error) {
 	values := parseQueries(transactions, url.Values{}, transactions.Usage, transactions.Log)
 
-	switch tokenType {
-	case "service_token":
-		values.Add("service_token", token)
-	case "provider_key":
-		values.Add("provider_key", token)
-	default:
-		return ApiResponse{}, errors.New("invalid token type value")
+	err := auth.SetURLValues(&values)
+	if err != nil {
+		return ApiResponse{}, err
```

Comment:
can we return here a `*ApiResponse` so that we can just `return nil, err` here?

Reply:
Right now the caller will check for err yes, but they can just call `resp.Success()` which will be false by default. Otherwise they would need, to avoid panic, check that `resp` is not nil before calling a function on it if we move to returning a pointer, no?

Reply:
Also in go, we return a pointer if we want the caller to be able to modify the contents, other than that, typical its preferable  to just return the value

Reply:
well, if they check `err` then they won't need to check the pointer as it would be valid - I think interfaces that return errors only to provide a escape hatches that would allow them to ignore the errors are dubious and prone to create strong demand for ibuprofen.

IMO such interfaces are useful whenever some computations are ok to "fail", because failing is not an error per-se but normal state of the program, or when you just do not or cannot care about it. Think for example about defining an iterator on an array: there is little point in calling it on an empty array, so you could return an error, but turns out that processing 0 elements is not an error, but as normal as processing 5 elements, and the end result still holds: "all elements of the array are processed".

Reply:
Yeah it is down to personal preference I guess. I don understand what your saying. If there is an error we arent expecting to do anything with the resp anyway so it will be thrown away regardless. My take on it was always asking the question, "Do we want/expect the caller to modify the value" if yes return a pointer.

I'm not overly concerned either way to be honest, if you think its better to change it then I have no major objections

Reply:
Sorry, perhaps my example was not the best in communicating what I see as a problem. Let me try to rephrase it.

My point is that when you do an API call you are very much interested in the result of that call and any errors that might occur. Using (from the caller) `resp, _ = client.ReportAppID(...); return resp.Success()` will _drop_ the error information, and I argue that that is wrong because you definitely care about the reason that the call failed - both to forward that information to some other place where it can be of help to anyone debugging, and to act on it if it is a transient, recoverable error, such as a limited network outage (ie. in which case we could schedule a new report with the required information).

Since you definitely care, then you will at some point need to check the `err` value, branching on the happy and not-so-happy paths, and in the case where `err == nil`, then `resp != nil`, so you can use the response object without panics.

The above is a reason to not ignore `err`, but is not a reason to pass a pointer. The reason to pass a pointer is that when you do not need to pass an `ApiResponse` object, you are paying the allocation price of a structure that has no purpose other than being freed when returning. IOW, you are going to use a whole lot of computation to signal something to the caller, even though something some orders of magnitude cheaper like a single bit could do, and in fact, you don't need to signal anything, since that's what the `err` object is for. Put another way, imagine whenever the method was successful we allocated a new `error` object just to say that there were no errors instead of returning `nil`!

This case also is irrespective to mutation, because regardless of pointers, the caller is able to mutate the returned `ApiResponse`, either directly or through the pointer. Perhaps you confused it with passing in a function parameter by pointer or by value? When passing it in by value the caller's object is not accessed, and instead a copy is passed in to the function, whereas with a pointer the callee can do whatever it pleases (almost) with the object pointed to.

Reply:
Sorry, I misunderstood your point about being able to call `Success()` on the response object and took it to mean you can just ignore errors.

Yes, I'd say the original concern is about avoiding a loss for no good enough reason, rather than just improving performance. :D